[MRG] Common Private Loss Module with tempita

lorentzenchr opened the pull request:
#### Reference Issues/PRs
Solves #15123.
This is an altenative to #19088. This PR uses Cython Tempita to generate Cython code from a template as is already done in `_sag_fast.pyx.tp` and `_seq_dataset.pyx.tp`.

See https://github.com/scikit-learn/scikit-learn/pull/19088#discussion_r634369849, issue #15123 and file `_loss.pyx.tp` (`# Design:`) for design choices.
Using Tempita helps reducing repeating code patterns for loops over samples in order to guarantee inlining.

lorentzenchr committed:
ENH add common link function submodule

lorentzenchr committed:
ENH add common link function submodule

lorentzenchr committed:
ENH add common loss function submodule

lorentzenchr committed:
ENH add common loss function submodule

lorentzenchr committed:
CLN replace deprecated np.int by int

lorentzenchr committed:
CLN replace deprecated np.int by int

lorentzenchr committed:
DOC document default=1 for n_threads

lorentzenchr committed:
DOC document default=1 for n_threads

lorentzenchr committed:
CLN comments and line wrapping

lorentzenchr committed:
CLN comments and line wrapping

lorentzenchr committed:
CLN comments and doc

lorentzenchr committed:
CLN comments and doc

lorentzenchr committed:
BUG remove useless line of code

lorentzenchr committed:
BUG remove useless line of code

lorentzenchr committed:
CLN remove line that was commented out

lorentzenchr committed:
CLN remove line that was commented out

lorentzenchr committed:
CLN nitpicks in comments and docstrings

lorentzenchr committed:
CLN nitpicks in comments and docstrings

lorentzenchr committed:
ENH set NPY_NO_DEPRECATED_API

lorentzenchr committed:
ENH set NPY_NO_DEPRECATED_API

lorentzenchr committed:
MNT change NPY_1_13_API_VERSION to NPY_1_7_API_VERSION

lorentzenchr committed:
MNT change NPY_1_13_API_VERSION to NPY_1_7_API_VERSION

lorentzenchr committed:
MNT comment out NPY_NO_DEPRECATED_API

lorentzenchr committed:
MNT comment out NPY_NO_DEPRECATED_API

lorentzenchr committed:
TST restructure domain test cases

lorentzenchr committed:
TST restructure domain test cases

lorentzenchr committed:
DOC add losses to API reference

lorentzenchr committed:
DOC add losses to API reference

lorentzenchr committed:
MNT add classes to __init__

lorentzenchr committed:
MNT add classes to __init__

lorentzenchr committed:
CLN fix import

lorentzenchr committed:
CLN fix import

lorentzenchr committed:
DOC minor docstring changes

lorentzenchr committed:
DOC minor docstring changes

lorentzenchr committed:
TST prefer docstring over comment

lorentzenchr committed:
TST prefer docstring over comment

lorentzenchr committed:
ENH define loss.is_multiclass

lorentzenchr committed:
ENH define loss.is_multiclass

lorentzenchr committed:
DOC fix typos

lorentzenchr committed:
DOC fix typos

lorentzenchr committed:
CLN address review comments

lorentzenchr committed:
CLN address review comments

lorentzenchr committed:
DOC small docstring improvements

lorentzenchr committed:
DOC small docstring improvements

lorentzenchr committed:
TST test more losses in test_specific_fit_intercept_only

lorentzenchr committed:
TST test more losses in test_specific_fit_intercept_only

lorentzenchr committed:
FIX test_loss_boundary

lorentzenchr committed:
FIX test_loss_boundary

lorentzenchr committed:
ENH Tempita for losses

lorentzenchr committed:
ENH Tempita for losses

lorentzenchr committed:
MNT apply black

lorentzenchr committed:
MNT apply black

lorentzenchr committed:
TST replace np.quantile by np.percentile

lorentzenchr committed:
TST replace np.quantile by np.percentile

ogrisel commented:
I suppose that we should get the same performance as in #19088 because in the end the generated code should be equivalent, right?

jjerphan commented:
> I suppose that we should get the same performance as in #19088 because in the end the generated code should be equivalent, right?

This should produce a nearly identical code, so I would think so.

lorentzenchr commented:
I can confirm that this PR (current state https://github.com/scikit-learn/scikit-learn/pull/20567/commits/98f88778c3ab0d40955a9ded83460955620e1116) is 100% identical to #19088 (except for a few blank lines in the pyx file).

ogrisel reviewed (approved):
I am just starting with the review but here are a few comments.

About the tempita, vs non tempita: this spares 600 lines of redundant code, so it's probably worth it but no strong feelings either.

ogrisel commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -0,0 +1,821 @@
+
+            h_numeric = numerical_derivative(grad_func, raw_prediction[:, k], eps=1e-6)
+            if loss.approx_hessian:
+                assert np.all(h >= h_numeric)
```
This is not covered either according to codecov.

lorentzenchr replied:
That's right. So far, no multi-class loss has an approximate hessian. The only multi-class loss is `CategoricalCrossEntropy` with exact (diagonal) hessian.

lorentzenchr replied:
There is scientific literature that uses such approximations, usually in the hope to make fitting faster or more robust.

One example is Fisher Scoring in MLE, e.g. in IRWLS for GLMs. There, instead of the Hessian=-observed information matrix, one uses the Fisher matrix=expected information matrix which is positive definite.

But I now realize that `h >= h_numeric` must not hold. For the Gamma deviance with log-link, we have
```
hessian = y_true * exp(-raw_prediction)
E[hessian] =  1
```
but `hessian>E[hessian]` is well possible. I'll better skip the test for `approx_hessian=True`!

lorentzenchr replied:
Removed in https://github.com/scikit-learn/scikit-learn/pull/20567/commits/7bee26b9dba71439af7a830d3a24753e586464bb with a TODO comment.

ogrisel commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -0,0 +1,821 @@
+    sp_version == parse_version("1.2.0"),
+    reason="bug in scipy 1.2.0, see scipy issue #9608",
+)
+@skip_if_32bit
```
Do you have any idea why this test is failing on 32 bit Python? I believe this is the same problem as #17073 but as far as I know the root cause is still not well understood.

I wonder which is platform dependent: our Cython loss function or something in the newton solver itself.

lorentzenchr replied:
Not really. Maybe, it is the use of C `double, exp, log` which are platform dependent in contrast to `npy_float64` which is always a 64-bit/8-byte IEEE 754 compatible floating point type.

ogrisel replied:
I think double is always a 64 bit float on all the platforms we support and test on a regular basis (x86, arm and ppcle, both 32 bit and 64 bit).

Not sure about `exp` and `log`.

jjerphan replied:
Which are the losses failing on 32 bit?

lorentzenchr replied:
This is a copy from the hgbt losses. We can investigate further, but I do not consider it a blocker for this PR.

ogrisel reviewed (commented):
Some more comments:

lorentzenchr commented:
> About the tempita, vs non tempita: this spares 600 lines of redundant code, so it's probably worth it but no strong feelings either.

As there are likely some more losses to come, e.g. Tweedie with identity link for GLM (which, today, I think was not the best idea), I meanwhile favor this tempita version. 

It might also make it easier in case we want better support 32bit loss versions.

> I am just starting with the review but here are a few comments.

@ogrisel Thank you very much for taking the time. Note that I won‘t be able to address your comments until mid of August.

lorentzenchr commented on sklearn/_loss/loss.py:
```diff
@@ -0,0 +1,910 @@
+            loss = np.empty_like(y_true)
+        # Be graceful to shape (n_samples, 1) -> (n_samples,)
+        if raw_prediction.ndim == 2 and raw_prediction.shape[1] == 1:
+            raw_prediction = raw_prediction.squeeze(1)
```
The use case is in HGBT
https://github.com/scikit-learn/scikit-learn/blob/df20e8156fdc06a89ba85952b8b5a32b47ee9004/sklearn/ensemble/_hist_gradient_boosting/loss.py#L184-L190
See also #19089 in gradient_boosting.py
https://github.com/scikit-learn/scikit-learn/blob/4e007384e7eb4e86553a1bb44a220d9a7ce48134/sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py#L522-L526

ogrisel replied:
Ok so this will be test once we do follow-up PRs to actually use this in HGBT then.

lorentzenchr replied:
I added a test for this lately in https://github.com/scikit-learn/scikit-learn/pull/20567/commits/f53e3f831ab08decf1776e0d4996ad85fa0b64c2.

jjerphan commented on sklearn/_loss/_loss.pyx.tp:
```diff
@@ -0,0 +1,1213 @@
+    with_param = ", self." + param
+}}
+
+cdef class {{name}}(cLossFunction):
```
I guess `cython.final` be useful here? From [the docs](https://cython.readthedocs.io/en/latest/src/tutorial/pure.html):

> `@cython.final` terminates the inheritance chain by preventing a type from being used as a base class, or a method from being overridden in subtypes. This enables certain optimisations such as inlined method calls.

From the code, it removes some boilerplate [for attribute lookups](https://github.com/cython/cython/blob/f6c850d11bc57a432519afb7369bf0d08008be25/Cython/Compiler/ModuleNode.py#L2315-L2319) and [for type safety checks](https://github.com/cython/cython/blob/f6c850d11bc57a432519afb7369bf0d08008be25/Cython/Compiler/ModuleNode.py#L1540-L1561). (Searching for [`is_final_type`](https://github.com/cython/cython/search?q=is_final_type) and [`is_final_cmethod`](https://github.com/cython/cython/search?q=is_final_cmethod) in Cython's code-base are good entry-points to see how this gets made).
```suggestion
@cython.final
cdef class {{name}}(cLossFunction):
```

jjerphan replied:
For instance, this is the diff in the generated C code when marking the following ext type as `final`:

```python
from cython cimport final
@final # change this
cdef class MyExtType:
    cdef int an_attribute
    @final # this is not necessary
    cdef int a_method(self, int an_argument):
        return 1

```
```patch
diff --git a/a.old.c b/a.c
index f32344844..f093c1be0 100644
--- a/a.old.c
+++ b/a.c
@@ -1040,10 +1040,10 @@ struct __pyx_obj_1a_MyExtType;
 
 /* "a.pyx":3
  * from cython cimport final
- * 
+ * @final
  * cdef class MyExtType:             # <<<<<<<<<<<<<<
  *     cdef int an_attribute
- * 
+ *     @final
  */
 struct __pyx_obj_1a_MyExtType {
   PyObject_HEAD
@@ -1057,6 +1057,7 @@ struct __pyx_vtabstruct_1a_MyExtType {
   int (*a_method)(struct __pyx_obj_1a_MyExtType *, int);
 };
 static struct __pyx_vtabstruct_1a_MyExtType *__pyx_vtabptr_1a_MyExtType;
+static int __pyx_f_1a_9MyExtType_a_method(struct __pyx_obj_1a_MyExtType *, int);
 /* #### Code section: utility_code_proto ### */
 
 /* --- Runtime support code (head) --- */
@@ -1411,13 +1412,6 @@ static CYTHON_INLINE PyObject* __Pyx_PyObject_GenericGetAttrNoDict(PyObject* obj
 #define __Pyx_PyObject_GenericGetAttrNoDict PyObject_GenericGetAttr
 #endif
 
-/* PyObject_GenericGetAttr.proto */
-#if CYTHON_USE_TYPE_SLOTS && CYTHON_USE_PYTYPE_LOOKUP && PY_VERSION_HEX < 0x03070000
-static PyObject* __Pyx_PyObject_GenericGetAttr(PyObject* obj, PyObject* attr_name);
-#else
-#define __Pyx_PyObject_GenericGetAttr PyObject_GenericGetAttr
-#endif
-
 /* SetVTable.proto */
 static int __Pyx_SetVtable(PyTypeObject* typeptr , void* vtable);
 
@@ -2039,7 +2033,7 @@ static int __pyx_m_traverse(PyObject *m, visitproc visit, void *arg) {
 
 /* "a.pyx":6
  *     cdef int an_attribute
- * 
+ *     @final
  *     cdef int a_method(self, int an_argument):             # <<<<<<<<<<<<<<
  *         return 1
  */
@@ -2050,7 +2044,7 @@ static int __pyx_f_1a_9MyExtType_a_method(CYTHON_UNUSED struct __pyx_obj_1a_MyEx
   __Pyx_RefNannySetupContext("a_method", 0);
 
   /* "a.pyx":7
- * 
+ *     @final
  *     cdef int a_method(self, int an_argument):
  *         return 1             # <<<<<<<<<<<<<<
  */
@@ -2059,7 +2053,7 @@ static int __pyx_f_1a_9MyExtType_a_method(CYTHON_UNUSED struct __pyx_obj_1a_MyEx
 
   /* "a.pyx":6
  *     cdef int an_attribute
- * 
+ *     @final
  *     cdef int a_method(self, int an_argument):             # <<<<<<<<<<<<<<
  *         return 1
  */
@@ -2865,11 +2859,7 @@ static PyObject *__pyx_tp_new_1a_MyExtType(PyTypeObject *t, CYTHON_UNUSED PyObje
   allocfunc alloc_func = (allocfunc)PyType_GetSlot(t, Py_tp_alloc);
   o = alloc_func(t, 0);
   #else
-  if (likely(!__Pyx_PyType_HasFeature(t, Py_TPFLAGS_IS_ABSTRACT))) {
-    o = (*t->tp_alloc)(t, 0);
-  } else {
-    o = (PyObject *) PyBaseObject_Type.tp_new(t, __pyx_empty_tuple, 0);
-  }
+  o = (*t->tp_alloc)(t, 0);
   if (unlikely(!o)) return 0;
   #endif
   p = ((struct __pyx_obj_1a_MyExtType *)o);
@@ -2878,11 +2868,6 @@ static PyObject *__pyx_tp_new_1a_MyExtType(PyTypeObject *t, CYTHON_UNUSED PyObje
 }
 
 static void __pyx_tp_dealloc_1a_MyExtType(PyObject *o) {
-  #if CYTHON_USE_TP_FINALIZE
-  if (unlikely((PY_VERSION_HEX >= 0x03080000 || __Pyx_PyType_HasFeature(Py_TYPE(o), Py_TPFLAGS_HAVE_FINALIZE)) && Py_TYPE(o)->tp_finalize) && (!PyType_IS_GC(Py_TYPE(o)) || !_PyGC_FINALIZED(o))) {
-    if (PyObject_CallFinalizerFromDealloc(o)) return;
-  }
-  #endif
   (*Py_TYPE(o)->tp_free)(o);
 }
 
@@ -2902,7 +2887,7 @@ static PyType_Spec __pyx_type_1a_MyExtType_spec = {
   "a.MyExtType",
   sizeof(struct __pyx_obj_1a_MyExtType),
   0,
-  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER|Py_TPFLAGS_BASETYPE,
+  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER,
   __pyx_type_1a_MyExtType_slots,
 };
 #else
@@ -2937,7 +2922,7 @@ static PyTypeObject __pyx_type_1a_MyExtType = {
   0, /*tp_getattro*/
   0, /*tp_setattro*/
   0, /*tp_as_buffer*/
-  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER|Py_TPFLAGS_BASETYPE, /*tp_flags*/
+  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER, /*tp_flags*/
   0, /*tp_doc*/
   0, /*tp_traverse*/
   0, /*tp_clear*/
@@ -3240,7 +3225,7 @@ static int __Pyx_modinit_type_init_code(void) {
   #endif
   #if !CYTHON_COMPILING_IN_LIMITED_API
   if ((CYTHON_USE_TYPE_SLOTS && CYTHON_USE_PYTYPE_LOOKUP) && likely(!__pyx_ptype_1a_MyExtType->tp_dictoffset && __pyx_ptype_1a_MyExtType->tp_getattro == PyObject_GenericGetAttr)) {
-    __pyx_ptype_1a_MyExtType->tp_getattro = __Pyx_PyObject_GenericGetAttr;
+    __pyx_ptype_1a_MyExtType->tp_getattro = __Pyx_PyObject_GenericGetAttrNoDict;
   }
   #endif
   if (__Pyx_SetVtable(__pyx_ptype_1a_MyExtType, __pyx_vtabptr_1a_MyExtType) < 0) __PYX_ERR(1, 3, __pyx_L1_error)
@@ -3586,7 +3571,7 @@ if (!__Pyx_RefNanny) {
 
   /* "a.pyx":1
  * from cython cimport final             # <<<<<<<<<<<<<<
- * 
+ * @final
  * cdef class MyExtType:
  */
   __pyx_t_1 = __Pyx_PyDict_NewPresized(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 1, __pyx_L1_error)
@@ -5231,16 +5216,6 @@ static CYTHON_INLINE PyObject* __Pyx_PyObject_GenericGetAttrNoDict(PyObject* obj
 }
 #endif
 
-/* PyObject_GenericGetAttr */
-#if CYTHON_USE_TYPE_SLOTS && CYTHON_USE_PYTYPE_LOOKUP && PY_VERSION_HEX < 0x03070000
-static PyObject* __Pyx_PyObject_GenericGetAttr(PyObject* obj, PyObject* attr_name) {
-    if (unlikely(Py_TYPE(obj)->tp_dictoffset)) {
-        return PyObject_GenericGetAttr(obj, attr_name);
-    }
-    return __Pyx_PyObject_GenericGetAttrNoDict(obj, attr_name);
-}
-#endif
-
 /* SetVTable */
 static int __Pyx_SetVtable(PyTypeObject *type, void *vtable) {
     PyObject *ob = PyCapsule_New(vtable, 0, 0);

```

lorentzenchr replied:
You could report those findings back to cython itself.

jjerphan reviewed (approved):
12b4634 (which is the addition from #19088) looks good to me.

lorentzenchr commented on sklearn/_loss/link.py:
```diff
@@ -0,0 +1,244 @@
+    if np.all(high):
+        return True
+    else:
+        return False
```
A dataclass with method `includes`. Good idea.

lorentzenchr committed:
ENH make Interval a dataclass

- function is_in_interval_range -> method Interval.includes

lorentzenchr committed:
ENH make Interval a dataclass

- function is_in_interval_range -> method Interval.includes

lorentzenchr committed:
DOC improve docstrings in link.py

lorentzenchr committed:
DOC improve docstrings in link.py

lorentzenchr committed:
MNT use numpy dtype instead of Python type

lorentzenchr committed:
MNT use numpy dtype instead of Python type

lorentzenchr commented on sklearn/_loss/link.py:
```diff
@@ -0,0 +1,244 @@
+        return False
+
+
+def _inclusive_low_high(interval, dtype=float):
```
With `np.float64` as default.

lorentzenchr replied:
> This function is just used for testing, right? If so, it should probably be moved in the test file itself.

Yes. But it is used in `test_link.py` and `test_loss.py`. Therefore, I placed it here.
I added a comment to the docstring that it is only used for testing.

ogrisel replied:
alright!

lorentzenchr committed:
TST add negative intervals

lorentzenchr committed:
TST add negative intervals

lorentzenchr committed:
ENH add __post_init__ to class Interval

lorentzenchr committed:
ENH add __post_init__ to class Interval

lorentzenchr committed:
MNT rename cython losses

- class name from cLoss to CyLoss
- single sample methods
  from closs to cy_loss, cgradient to cy_gradient, ..

lorentzenchr committed:
MNT rename cython losses

- class name from cLoss to CyLoss
- single sample methods
  from closs to cy_loss, cgradient to cy_gradient, ..

lorentzenchr commented on sklearn/_loss/_loss.pxd:
```diff
@@ -0,0 +1,75 @@
+
+
+# C base class for loss functions
+cdef class cLossFunction:
```
Done. Do you think the python method names with leading underscore, e.g. `CyLoss._loss` are still fine?

lorentzenchr committed:
TST loss.predict_proba

lorentzenchr committed:
TST loss.predict_proba

lorentzenchr committed:
TST predict_proba and gradient_proba

lorentzenchr committed:
TST predict_proba and gradient_proba

lorentzenchr committed:
MNT use is_multiclass in tests instead of n_classes <= 2

lorentzenchr committed:
MNT use is_multiclass in tests instead of n_classes <= 2

lorentzenchr committed:
DOC docstring predict_proba and more

lorentzenchr committed:
DOC docstring predict_proba and more

lorentzenchr committed:
Merge branch 'main' into loss_module_tempita

lorentzenchr committed:
MNT remove top_path from gen_from_templates

lorentzenchr committed:
MNT remove top_path from gen_from_templates

lorentzenchr committed:
CI add --allow-releaseinfo-change in circleci

lorentzenchr commented:
ci/circleci: doc got the following error in `build_doc.sh`:
```
+ sudo -E apt-get -yq update
Get:1 http://security.debian.org/debian-security buster/updates InRelease [65.4 kB]
Get:2 http://deb.debian.org/debian buster InRelease [122 kB]
Get:3 http://deb.debian.org/debian buster-updates InRelease [51.9 kB]
Reading package lists...
E: Repository 'http://security.debian.org/debian-security buster/updates InRelease' changed its 'Suite' value from 'stable' to 'oldstable'
E: Repository 'http://deb.debian.org/debian buster InRelease' changed its 'Suite' value from 'stable' to 'oldstable'
E: Repository 'http://deb.debian.org/debian buster-updates InRelease' changed its 'Suite' value from 'stable-updates' to 'oldstable-updates'
```
Adding the option `--allow-releaseinfo-change` in https://github.com/scikit-learn/scikit-learn/pull/20567/commits/5721077dbaf430b25ba9d7a1516491066189d945 solved this issue.
This issue might warrant a separate (fast) CI fix PR.

lorentzenchr commented:
Some remaining questions:
- Will the class methods on single sample points like `cdef double cy_loss` be inlined in SGD (`_sgd_fast.pyx`)?
- Should we keep the names `BinaryCrossEntropy` and `CategoricalCrossEntropy`? Or use other names like `HalfBinomialDeviance`, `HalfMultinomialDeviance` or `LogisticLoss`, `BinaryLogLoss`, `MultinomialLogLoss`?

lorentzenchr committed:
TST test graceful squeezing

lorentzenchr committed:
TST test graceful squeezing

lorentzenchr committed:
Merge branch 'main' into loss_module_tempita

lorentzenchr committed:
CLN no extra methods for HalfSquaredError

lorentzenchr committed:
CLN no extra methods for HalfSquaredError

lorentzenchr commented:
CI finally green (codecov was the only red one before).

ogrisel commented:
> Should we keep the names BinaryCrossEntropy and CategoricalCrossEntropy? Or use other names like HalfBinomialDeviance, HalfMultinomialDeviance or LogisticLoss, BinaryLogLoss, MultinomialLogLoss?

I would be in favor of keeping `BinaryCrossEntropy` and `CategoricalCrossEntropy` as they are already used in popular Deep Learning frameworks, and mention alternative names in the docstrings both of the private API and the public estimator that use those loss functions.

ogrisel commented:
> Will the class methods on single sample points like cdef double cy_loss be inlined in SGD (_sgd_fast.pyx)?

I don't understand this question, could you please rephrase or refine?

ogrisel reviewed (approved):
I confirm that I am ok with merging this PR as it is. Since it's private only, it should not impact the users in anyway. But then it will allow us to review the subsequent PRs that actually enable this in public estimators. We can always refine the implementation / API in those PRs since this is all private.

lorentzenchr commented:
> Will the class methods on single sample points like cdef double cy_loss be inlined in SGD (_sgd_fast.pyx)?

Currently, SGD uses its own extension type loss classes like
```cython
cdef class SquaredLoss(Regression):
    cdef double loss(self, double p, double y) nogil
    cdef double dloss(self, double p, double y) nogil
```
Then they  are called in
```cython
def _plain_sgd(..., LossFunction loss,...)
    ...
    for i in range(n_samples):
        sumloss += loss.loss(p, y)  # function called inlined?
        ...
        dloss = loss.dloss(p, y)    # function called inlined?
```
My question is: Will the call to `loss.loss(p, y)` be inlined either by Cython (generating C code) or by the C compiler?
If so, would this still be the case with the new losses of this PR?

*My attempt at an answer:*
The currently generated C file reads:
```C
__pyx_v_sumloss = (__pyx_v_sumloss + ((struct __pyx_vtabstruct_7sklearn_12linear_model_9_sgd_fast_LossFunction *)__pyx_v_loss->__pyx_vtab)->loss(__pyx_v_loss, __pyx_v_p, __pyx_v_y));
```
So I think even today it is not inlined and there are no worries.

lorentzenchr committed:
TST remove testing if approx_hessian=True

lorentzenchr committed:
TST remove testing if approx_hessian=True

lorentzenchr commented on sklearn/_loss/_loss.pyx.tp:
```diff
@@ -1,3 +1,164 @@
+    Link:
+    y_pred = exp(raw_prediction)
+
+    Half Tweedie deviance with log-link and p=power is
+        max(y_true, 0)**(2-p) / (1-p) / (2-p)
+        - y_true * y_pred**(1-p) / (1-p)
+        + y_pred**(2-p) / (2-p)
+        = max(y_true, 0)**(2-p) / (1-p) / (2-p)
+        - y_true * exp((1-p) * raw_prediction) / (1-p)
+        + exp((2-p) * raw_prediction) / (2-p)
+
+    Dropping constant terms, this gives:
+        exp((2-p) * raw_prediction) / (2-p)
+        - y_true * exp((1-p) * raw_prediction) / (1-p)
```
@glemaitre You seem to be to docstring master? Do you have an advice?

jjerphan replied:
[`DistanceMetric`'s expressions are present in their doc-strings using this directive](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/neighbors/_dist_metrics.pyx#L419-L420), yet their aren't publicly documented.

lorentzenchr replied:
I find latex on not rendered docstrings less readable.

lorentzenchr commented on doc/modules/classes.rst:
```diff
@@ -1646,3 +1646,27 @@ Recently deprecated
 
 To be removed in 1.0 (renaming of 0.25)
 ---------------------------------------
+
+.. _loss_function_ref:
+
+:mod:`sklearn._loss`: Private Loss Function Classes
+===========================================================
+
```
Transfer of discussion https://github.com/scikit-learn/scikit-learn/pull/20811#discussion_r693728993:
Quoting @NicolasHug 
> Is there a specific reason to document these?
>
>I'm not sure we should: we never expose private stuff in the API ref so far, and even though it's clear from here that these are private, it won't be as clear once users land on the individual pages of such losses. Until we have a system to visually distinguish the private stuff from the public stuff in the docs, I would recommend not to document them publicly. (related is our developer-API discussion)

@ogrisel
> Maybe we could move this to a dedicated page, somewhere under the developer documentation sections of the documentation.

NicolasHug replied:
What developer docs are you referring to? IMHO these don't even qualify as relevant developer utils.

lorentzenchr replied:
See also https://github.com/scikit-learn/scikit-learn/pull/19088#issuecomment-812451036:
> Based on @cmarmo's suggestion in https://github.com/scikit-learn/scikit-learn/pull/19088#pullrequestreview-625124432, I've the same question: To render or not to render docstrings of this private module under API Reference?

NicolasHug replied:
thanks for the refs. Let's remove them please. We don't document private implementation details, there's no reason to start now.

lorentzenchr replied:
Alright.

lorentzenchr committed:
DOC remove loss module for classes.rst

lorentzenchr committed:
DOC remove loss module for classes.rst

lorentzenchr committed:
TST that losses can be pickled

lorentzenchr committed:
TST that losses can be pickled

da-woods commented:
> **Quote:** "I'm not quite sure what the solution is at this stage, but I recommend not inheriting from multiple cdef classes!"
> **This means the class inheritance design has to be reworked!**

See https://github.com/cython/cython/issues/4350#issuecomment-907604319 for full details. I think it's OK provided that that the most derived Cython class should end up in C.__base__. So you may be OK just re-ordering a few bases.

lorentzenchr committed:
TST add test_loss_on_specific_values

lorentzenchr committed:
TST add test_loss_on_specific_values

lorentzenchr committed:
FIX make cython inheritance happy and losses pickable

lorentzenchr committed:
FIX make cython inheritance happy and losses pickable

lorentzenchr commented:
@ogrisel @jjerphan You might want to have a look at the changed order of inheritance in https://github.com/scikit-learn/scikit-learn/pull/20567/commits/cef6e24cc8a12a1a9605da06d7462cbd8fc8e5c1 and maybe approve againt if you are OK with it (CI is all green).

To make this PR bullet proof, losses are explicitly checked against manual computations in https://github.com/scikit-learn/scikit-learn/pull/20567/commits/696d18fe39a011d2c90dc876f14bade1bd6789f7, and https://github.com/scikit-learn/scikit-learn/pull/20567/commits/67a30693e81a37e8ba5e0f8353ff1b56ed879bac introduced checks that losses are pickable (which revealed the problems mentioned in https://github.com/scikit-learn/scikit-learn/pull/20567#issuecomment-907289384)

lorentzenchr commented:
A question on readonly inputs: As long as we wait for Cython 3.0 release, i.e. for support of const memoryviews with fused types, how should the losses tackle this. They use fused types without `const` qualifier and can thus not handle readonly arrays like memory mapped arrays.

As a consequence in #20811, `test_estimators[HistGradientBoostingRegressor()-check_regressors_train(readonly_memmap=True)]` fails with `E   ValueError: buffer source array is read-only`. 
[For HGBT, this could be solved by copying `y` in the readonly case (which might be ok as it is much cheaper compared to copying `X`). I'm not so sure how to solve this for other estimators.]

da-woods commented:
> A question on readonly inputs: As long as we wait for Cython 3.0 release, i.e. for support of const memoryviews with fused types, how should the losses tackle this. They use fused types without `const` qualifier and can thus not handle readonly arrays like memory mapped arrays.

If you genuinely treat the arrays as readonly then you could always mislead Python with a wrapper class that marks them as writeable:

```
from cpython cimport Py_buffer
from cpython.buffer cimport PyObject_GetBuffer, PyBuffer_Release, PyBUF_WRITABLE

cdef class ReadonlyWrapper:
    cdef object wraps

    def __init__(self, wraps):
        self.wraps = wraps

    def __getbuffer__(self, Py_buffer *buffer, int flags):
        request_for_writeable = False
        if flags & PyBUF_WRITABLE:
            flags ^= PyBUF_WRITABLE
            request_for_writeable = True
        PyObject_GetBuffer(self.wraps, buffer, flags)
        if request_for_writeable:
            buffer.readonly = False  #lie!

    def __releasebuffer__(self, Py_buffer *buffer):
        PyBuffer_Release(buffer)
```

This class supports the buffer protocol (so can be passed to a memoryview) but always reports the contents as writeable.

Obviously if you do modify one of these fake "writeable" arrays then you're probably in a big mess. So use it very carefully and treat it like the hack that it is. And delete it as soon as you move to Cython 3.

lorentzenchr committed:
ENH support const memoryviews by ReadonlyWrapper

lorentzenchr committed:
ENH support const memoryviews by ReadonlyWrapper

lorentzenchr commented:
@da-woods This is so great! My headaches are gone:smirk: `ReadonlyWrapper` works as expected.

For reviewers, i.e. @ogrisel @jjerphan, I added this in https://github.com/scikit-learn/scikit-learn/pull/20567/commits/130306c5bf69708f5785cfd8d338b894c84ceb0a with additional tests to check readonly inputs already for losses (not only test_common.py).

jjerphan commented:
Thanks @da-woods for your help!

`ReadonlyWrapper` can also be used in other places of the code-base which met this issue, such as [in KMeans](https://github.com/scikit-learn/scikit-learn/blob/8c4589b23c6481f978d4cfab511f25b77a805f13/sklearn/cluster/_k_means_lloyd.pyx#L5-L7).

jjerphan commented on sklearn/_loss/loss.py:
```diff
@@ -434,7 +434,11 @@ def constant_to_optimal_zero(self, y_true, sample_weight=None):
 
 
-class HalfSquaredError(IdentityLink, BaseLoss, CyHalfSquaredError):
+# Note: Naturally, we would inherit in the following order
+#         class HalfSquaredError(IdentityLink, CyHalfSquaredError, BaseLoss)
+#       But because of https://github.com/cython/cython/issues/4350 we
+#       set BaseLoss as the last one. This, of course, changes the MRO.
```
Thanks for the comment.

lorentzenchr replied:
de rien:smile:

jjerphan commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -237,37 +237,48 @@ def test_loss_on_specific_values(loss, y_true, raw_prediction, loss_true):
 @pytest.mark.parametrize("n_threads", [1, 2])
-def test_loss_dtype(loss, dtype_in, dtype_out, sample_weight, out1, out2, n_threads):
-    """Test acceptance of dtypes in loss functions.
+def test_loss_dtype_readonly(
+    loss, readonly_memmap, dtype_in, dtype_out, sample_weight, out1, out2, n_threads
+):
+    """Test acceptance of dtypes and readonly arrays in loss functions.
```
As this test for both readonly and write-able arrays, this patch on the test name and top of the doc-string might be reverted.

lorentzenchr replied:
What exactly do you propose?

jjerphan replied:
Something like:
```suggestion
def test_loss_dtype(
    loss, readonly_memmap, dtype_in, dtype_out, sample_weight, out1, out2, n_threads
):
    """Test acceptance of dtypes and readonly and write-able arrays in loss functions.
```

This is more of a nitpick that you can ignore.

jjerphan reviewed (commented):
Some minor comments and suggestions on new commits since my last review (3f8ffe9 to 130306c5bf69708f5785cfd8d338b894c84ceb0a).

Apart from those this LGTM.

lorentzenchr commented:
> Thanks @da-woods for your help!
> 
> `ReadonlyWrapper` can also be used in other places of the code-base which met this issue, such as [in KMeans](https://github.com/scikit-learn/scikit-learn/blob/8c4589b23c6481f978d4cfab511f25b77a805f13/sklearn/cluster/_k_means_lloyd.pyx#L5-L7).

Should we then place it somewhere in `utils`? E.g. in `arrayfuncs.pyx` or a new file `_readonly_array_wrapper.pyx`?

lorentzenchr commented on sklearn/_loss/_loss.pxd:
```diff
@@ -0,0 +1,75 @@
+
+
+cdef class CyPinballLoss(CyLossFunction):
+    cdef readonly double quantile  # readonly makes it inherited by children
```
Yes. From [Cython documentation](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html?highlight=readable#static-attributes):
> By default, extension type attributes are only accessible by direct access, not Python access, which means that they are not accessible from Python code. To make them accessible from Python code, you need to declare them as **public** or **readonly**. For example:

I'll change the comment.

jjerphan commented:
If it's temporary and private, probably a new file (with the naming you suggest) is appropriate.

Also, it's probably better to create a PR introducing this fixture first in the code base and to then adapt this PR to use it. What do you think?

lorentzenchr commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -468,6 +488,37 @@ def test_sample_weight_multiplies_gradients(loss, sample_weight):
+    )
+
+    if raw_prediction.ndim == 1:
+        raw_prediction_2d = raw_prediction[:, None]
```
The Cython functions like `_loss` only work for `raw_prediction.shape = (n_samples,)`, but HGBT, for example, has `raw_prediction.shape = (n_samples, 1)`. That's why `def loss` and others have:
```python
if raw_prediction.ndim == 2 and raw_prediction.shape[1] == 1:
            raw_prediction = raw_prediction.squeeze(1)
```
I changed the docstring of this test.

jjerphan replied:
LGTM.

lorentzenchr commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -916,3 +918,21 @@ def test_predict_proba(loss):
+    )
+    pickled_loss = pickle.dumps(loss)
+    unpickled_loss = pickle.loads(pickled_loss)
+    assert loss(y_true=y_true, raw_prediction=raw_prediction) == approx(
+        unpickled_loss(y_true=y_true, raw_prediction=raw_prediction)
+    )
```
I copied the test from `check_estimators_pickle`. I guess there could be some issues depending on the platform? But I don't know the real reasons, TBH.

jjerphan replied:
Probably some pickle technical details? I believe it's fine keeping it as is.

lorentzenchr commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -190,6 +196,46 @@ def test_loss_boundary_y_pred(loss, y_pred_success, y_pred_fail):
+)
+def test_loss_on_specific_values(loss, y_true, raw_prediction, loss_true):
+    """Test losses at specific values."""
+    assert loss(
+        y_true=np.array([y_true]), raw_prediction=np.array([raw_prediction])
+    ) == approx(loss_true)
```
+1 for tolerances

lorentzenchr committed:
address review comments

lorentzenchr committed:
address review comments

jjerphan commented on sklearn/_loss/tests/test_loss.py:
```diff
@@ -609,7 +609,8 @@ def grad_func(x):
         h_numeric = numerical_derivative(grad_func, raw_prediction, eps=1e-6)
         if loss.approx_hessian:
-            assert np.all(h >= h_numeric)
+            # TODO: What could we test if loss.approx_hessian?
```
Would testing for numerical closeness with larger tolerances qualify?

lorentzenchr replied:
Unfortunately no. I guess we have to wait for concrete cases.

jjerphan reviewed (approved):
Apart from the (yet unresolved ?) [discussion regarding failing losses on 32 bit](https://github.com/scikit-learn/scikit-learn/pull/20567#discussion_r678066812) and the eventual prior introduction of `ReadonlyWrapper` in the code-base, this LGTM!

lorentzenchr committed:
FEA add ReadonlyWrapper

lorentzenchr committed:
CLN nitpick

lorentzenchr committed:
CLN import ReadonlyWrapper from utils

lorentzenchr committed:
Merge

lorentzenchr commented:
I rebased on #20903, https://github.com/scikit-learn/scikit-learn/pull/20903/commits/86d659cea3c6d11bcd9b423c8774f6d5f62b592f and remove the `ReadonlyWrapper` from this PR in https://github.com/scikit-learn/scikit-learn/pull/20567/commits/2820be5eabf5725c972cbbc68e6a4e4023489cc6. I think it would make sense to first merge the `ReadonlyWrapper` and then this PR.

lorentzenchr committed:
Merge branch 'main' into loss_module_tempita

lorentzenchr committed:
MNT replace ReadonlyWrapper by ReadonlyArrayWrapper

jjerphan reviewed (approved):
LGTM now that `ReadonlyArrayWrapper` got in.

Thank you @lorentzenchr!

lorentzenchr committed:
trigger CI

lorentzenchr commented:
Is there anything left or are we good to go?

jjerphan commented:
Yes, this PR just needs another maintainer approval.

lorentzenchr commented:
I'd say we have 2 very thorough reviews, and it might just need a core dev with the courage to press this inconspicuous green button 😄

jjerphan commented:
> I'd say we have 2 very thorough reviews, and it might just need a core dev with the courage to press this inconspicuous green button smile

I also believe that this PR needs love.

@NicolasHug: as you commented this PR, would you be interested in doing a review? :)

NicolasHug commented:
> @NicolasHug: as you commented this PR, would you be interested in doing a review? :)

Sorry, I won't have time for this :/

thomasjpfan commented:
I will review

thomasjpfan reviewed (commented):
First pass over the Python implementation

lorentzenchr committed:
MNT rename out parameters

For instance, loss -> loss_out and so on.

lorentzenchr committed:
CLN address review comments

lorentzenchr committed:
TST increase maxiter

lorentzenchr committed:
Merge branch 'main' into loss_module_tempita

lorentzenchr committed:
Revert "TST increase maxiter"

This reverts commit 5f3b0a5f8a0a73baef82c4f969b064d501731f1d.

lorentzenchr committed:
MNT composition instead of inheritance

lorentzenchr committed:
MNT interval_raw_prediction never used

lorentzenchr commented on sklearn/_loss/loss.py:
```diff
@@ -0,0 +1,912 @@
+    differentiable = True
+    is_multiclass = False
+
+    def __init__(self, n_classes=1):
```
Done. No more inheritance but composition.

thomasjpfan commented on sklearn/_loss/loss.py:
```diff
@@ -0,0 +1,923 @@
+    differentiable = True
+    is_multiclass = False
+
+    def __init__(self, n_classes=1):
```
Since the bass class uses them directly, I think it's better to set these attributes here and have child classes pass them up:

```suggestion
    def __init__(self, closs, link, n_classes=1):
        self.closs = closs
        self.link = link
```

lorentzenchr replied:
Good point.

lorentzenchr committed:
CLN closs and link as args in __init__

lorentzenchr committed:
trigger CI

lorentzenchr commented on sklearn/_loss/loss.py:
```diff
@@ -0,0 +1,912 @@
+        self.interval_y_true = Interval(-np.inf, np.inf, False, False)
+        self.interval_y_pred = Interval(-np.inf, np.inf, False, False)
+
+    def in_y_true_range(self, y):
+        """Return True if y is in the valid range of y_true.
+
+        Parameters
+        ----------
+        y : ndarray
+        """
+        return self.interval_y_true.includes(y)
```
I went with composition, see https://github.com/scikit-learn/scikit-learn/pull/20567#discussion_r734081815. Thus we have attributes `closs` and `link`. This gives us `loss_obj.link.inverse()`, which seems clear enough for me (instead of `loss_obj.link.inverse_link()`).

lorentzenchr commented:
I have no idea bout the CI failures. But I would love to get this PR in this year.

jjerphan commented:
> I have no idea bout the CI failures.

Neither do I. I can't reproduce the problem on my machine.
Could you create an empty commit to trigger the CI again?

> But I would love to get this PR in this year.

So do I. We will get there soon. :slightly_smiling_face:

lorentzenchr committed:
trigger CI

lorentzenchr committed:
Merge branch 'main' into loss_module_tempita

lorentzenchr committed:
DEBUG print infos

lorentzenchr commented:
The CI error on Linux_Docker pypy3 disappeared miraculously. But the segmentation fault in Ubuntu_Bionic py37_conda_forge_openblas_ubuntu_1804 in `test_loss_dtype` first call  of `loss.loss(..)` remains. I've not the faintest idea. Interestingly, it is not the first test that calls `loss.loss()`, but the first that fails.

This CI failure did not happen before. I don't know if it's worth to dig in deeper. Unfortunately, it's hard to get any debug information with a segmentation fault, grrrr.

jjerphan commented:
Good luck, @lorentzenchr!

jjerphan commented:
I will try to reproduce it other machines than mine.

In the meantime, good luck, @lorentzenchr!

jjerphan commented:
If this can help, a test fails on one machine:

```
__________________________________________________________________________ test_quantile_estimates_calibration[0.9] __________________________________________________________________________
[gw0] linux -- Python 3.9.5 /home/jjerphan/.local/share/miniconda3/envs/sk/bin/python

q = 0.9

    @pytest.mark.parametrize("q", [0.5, 0.9, 0.05])
    def test_quantile_estimates_calibration(q):
        # Test that model estimates percentage of points below the prediction
        X, y = make_regression(n_samples=1000, n_features=20, random_state=0, noise=1.0)
        quant = QuantileRegressor(
            quantile=q,
            alpha=0,
            solver_options={"lstsq": False},
        ).fit(X, y)
>       assert np.mean(y < quant.predict(X)) == approx(q, abs=1e-2)
E       AssertionError: assert 0.721 == 0.9 ± 1.0e-02
E        +  where 0.721 = <function mean at 0x7f9350752b80>(array([ 3.30604736e+02, -3.23803566e+01,  1.72624077e+02,  2.30198732e+02,\n        8.65493133e+01, -2.62029465e+00, -4...4218e+01, -5.61245674e+01,  3.39451460e+01,\n       -3.69123822e+02, -3.30180436e+02,  9.61597903e+01, -7.96443511e+01]) < array([ 3.34483269e+02, -2.89499437e+01,  1.73470656e+02,  2.31449580e+02,\n        8.85535349e+01, -2.58031839e+00, -4...8727e+01, -5.56012945e+01,  3.70361168e+01,\n       -3.68963303e+02, -3.28440336e+02,  9.63252109e+01, -8.09209754e+01]))
E        +    where <function mean at 0x7f9350752b80> = np.mean
E        +    and   array([ 3.34483269e+02, -2.89499437e+01,  1.73470656e+02,  2.31449580e+02,\n        8.85535349e+01, -2.58031839e+00, -4...8727e+01, -5.56012945e+01,  3.70361168e+01,\n       -3.68963303e+02, -3.28440336e+02,  9.63252109e+01, -8.09209754e+01]) = <bound method LinearModel.predict of QuantileRegressor(alpha=0, quantile=0.9, solver_options={'lstsq': False})>(array([[-0.98166842, -0.26256223,  0.84595188, ...,  0.13100719,\n         0.04944784,  0.23461727],\n       [-0.2535520...727,  1.25129233],\n       [-0.17643854, -0.01664586,  0.79574082, ...,  0.85314816,\n        -1.63343453,  0.14376031]]))
E        +      where <bound method LinearModel.predict of QuantileRegressor(alpha=0, quantile=0.9, solver_options={'lstsq': False})> = QuantileRegressor(alpha=0, quantile=0.9, solver_options={'lstsq': False}).predict
E        +  and   0.9 ± 1.0e-02 = approx(0.9, abs=0.01)

X          = array([[-0.98166842, -0.26256223,  0.84595188, ...,  0.13100719,
         0.04944784,  0.23461727],
       [-0.2535520...727,  1.25129233],
       [-0.17643854, -0.01664586,  0.79574082, ...,  0.85314816,
        -1.63343453,  0.14376031]])
q          = 0.9
quant      = QuantileRegressor(alpha=0, quantile=0.9, solver_options={'lstsq': False})
y          = array([ 3.30604736e+02, -3.23803566e+01,  1.72624077e+02,  2.30198732e+02,
        8.65493133e+01, -2.62029465e+00, -4...4218e+01, -5.61245674e+01,  3.39451460e+01,
       -3.69123822e+02, -3.30180436e+02,  9.61597903e+01, -7.96443511e+01])

sklearn/linear_model/tests/test_quantile.py:112: AssertionError
```

<details>
<summary>Installation versions</summary>

```
python -c "import sklearn; sklearn.show_versions()"

System:
    python: 3.9.5 (default, May 18 2021, 19:34:48)  [GCC 7.3.0]
executable: /home/jjerphan/.local/share/miniconda3/envs/sk/bin/python
   machine: Linux-5.14.15-200.fc34.x86_64-x86_64-with-glibc2.33

Python dependencies:
          pip: 21.3.1
   setuptools: 58.5.3
      sklearn: 1.1.dev0
        numpy: 1.20.3
        scipy: 1.7.2
       Cython: 0.29.24
       pandas: 1.3.4
   matplotlib: 3.4.3
       joblib: 1.1.0
threadpoolctl: 3.0.0

Built with OpenMP: True

threadpoolctl info:
       user_api: openmp
   internal_api: openmp
         prefix: libgomp
       filepath: /home/jjerphan/.local/share/miniconda3/envs/sk/lib/libgomp.so.1.0.0
        version: None
    num_threads: 8

       user_api: blas
   internal_api: openblas
         prefix: libopenblas
       filepath: /home/jjerphan/.local/share/miniconda3/envs/sk/lib/libopenblasp-r0.3.17.so
        version: 0.3.17
threading_layer: pthreads
   architecture: SkylakeX
    num_threads: 8

       user_api: blas
   internal_api: openblas
         prefix: libopenblas
       filepath: /home/jjerphan/.local/share/miniconda3/envs/sk/lib/python3.9/site-packages/scipy.libs/libopenblasp-r0-8b9e111f.3.17.so
        version: 0.3.17
threading_layer: pthreads
   architecture: SkylakeX
    num_threads: 8
```


</details>

lorentzenchr commented:
If I decipher the traceback correctly, seg faults are in:
- `_loss/tests/test_loss.py::test_loss_dtype[1-None-None-None-float32-float32-True-HalfGammaLoss] Fatal Python error: Segmentation fault`
- `_loss/tests/test_loss.py::test_loss_dtype[1-None-1-None-float32-float64-True-HalfSquaredError] Fatal Python error: Segmentation fault`

But those could also be at random.

lorentzenchr commented:
@jjerphan Thanks for your support. That really means a lot to me.
I'm hunting the segmentation fault in `test_loss_dtype` which only happens in the `Ubuntu_Bionic` CI run (e.g. see [this CI run](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=34950&view=logs&j=4371db68-60d2-5800-021d-f190a09db133&t=ae1b5267-f8ef-579b-7034-78324f7c3041&l=648)).

lorentzenchr committed:
DEBUG remove readonly_memmap test

lorentzenchr commented:
As https://github.com/scikit-learn/scikit-learn/pull/20567/commits/903be0be4c85090d7b97dc3b1a4d2e8965413251 passes the tests, the failure seems to be caused when data from `create_memmap_backed_data` is passed to the losses. It may be related to the `ReadonlyArrayWrapper`. It's hard to tell where exactly the failure happens.

lorentzenchr committed:
Revert "DEBUG remove readonly_memmap test"

This reverts commit 903be0be4c85090d7b97dc3b1a4d2e8965413251.

TomDLT commented:
It might be related to misaligned arrays in joblib (https://github.com/joblib/joblib/issues/563).

Numpy checks for misaligned data in every function, so it is not a problem in most estimators, but C expects arrays to be aligned to a multiple of their size. You code seems to directly calls C code, hence the segfault.

```py
import numpy as np

def create_memmap_backed_data(data):
    import os.path
    import tempfile
    import joblib
    temp_folder = tempfile.mkdtemp(prefix='testing_')
    filename = os.path.join(temp_folder, 'data.pkl')
    joblib.dump(data, filename)
    return joblib.load(filename, mmap_mode='r')

y = create_memmap_backed_data(np.random.randn(20, 1))
print(y.flags['ALIGNED'])
# If this is False for some reason, using C code on y will segfault.
```

lorentzenchr commented:
@TomDLT Thanks for helping. If your guess is correct, do you have an explanation why it only happens on `Ubuntu_Bionic py37_conda_forge_openblas_ubuntu_1804`?

lorentzenchr committed:
TST skip test if data not aligned

lorentzenchr committed:
Revert "DEBUG print infos"

This reverts commit 0068c684fce8119f87dba1a77e066643733754a5.

lorentzenchr commented:
@TomDLT What do you propose as a solution?
I see the following options:

1. Remove the readonly tests
2. Use [numpy.memmap](https://numpy.org/doc/stable/reference/generated/numpy.memmap.html) instead

More infos:
```python
y = create_memmap_backed_data(np.random.randn(20, 1))
print(y.flags['ALIGNED'])
```
is always `False`. This means that skipping the test if data is not aligned, https://github.com/scikit-learn/scikit-learn/pull/20567/commits/ef3b9e79de161b02d12aba64eb922c5479d595d1, always skips the tests for memory mapped data.

Cython `x[::1]` seems to only check if the array C/F-contigous, but not if it is aligned. I still don't get it, why this is a problem only for `Ubuntu_Bionic py37_conda_forge_openblas_ubuntu_1804`.

lorentzenchr committed:
TST zeros instead of empty initial guess

jjerphan commented:
@lorentzenchr: as all the CI jobs now pass, do you still need help for inspection? Are your last modifications final?

lorentzenchr commented:
@jjerphan My modifications are not final. https://github.com/scikit-learn/scikit-learn/pull/20567/commits/ef3b9e79de161b02d12aba64eb922c5479d595d1 skips all tests with memory mapped arrays as joblib always gives back `y.flags['ALIGNED'] = False` (on all platforms, I think).
So the error comes from the mix of joblib for memory mapped readonly arrays together with contiguous memoryviews in Cython, and only on Ubuntu 1804.
Questions are then:

1. Is this a problem of this PR or can we ignore it? We don't know the deeper reason for it.
2. Should we disable readonly arrays for the losses as long as we can't fix this test?
3. How to proceed with the test? Just remove the part with the readonly array? Or use numpy.memmap directly which might work.

jjerphan commented:
I think solving it by first isolating and identifying the problem with joblib (hence relating to 1.) would be better because the current tests are tailored for existent read-only arrays usages done with or from this library.

lorentzenchr committed:
Revert "TST skip test if data not aligned"

This reverts commit ef3b9e79de161b02d12aba64eb922c5479d595d1.

lorentzenchr committed:
DEBUG set boundscheck=True

lorentzenchr commented:
Side note: Setting `# cython: boundscheck=True` in the `_loss.pyx.tp` file did not have any effect, but setting `"boundscheck": True,` in the `sklear/build_utils/__init__.py` did.

lorentzenchr commented:
The findings in https://github.com/scikit-learn/scikit-learn/pull/21654#issuecomment-968349075 showed that the segfault is caused by https://github.com/joblib/joblib/issues/563. So it's not an issue with this PR.

lorentzenchr committed:
Revert "DEBUG set boundscheck=True"

This reverts commit e9c551c0604b43092ae5a94ea08624cb50c4cbcc.

lorentzenchr committed:
CLN setup.py

lorentzenchr committed:
CLN rename binary and categorical cross entropy to binomial and multinomial loss

lorentzenchr commented:
For reviewers: I dared to rename the classification losses in https://github.com/scikit-learn/scikit-learn/pull/20567/commits/00328b29ed5b8192ea7bccb9a1550e8b45fee9f4
- `BinaryCrossEntropy`-> `HalfBinomialLoss`
- `CategoricalCrossEntropy` -> `HalfMultinomialLoss`

I think this makes the names more consistent. If you dislike it, it's easy to revert that commit.

ogrisel commented:
> I dared to rename the classification losses in 00328b2
>    BinaryCrossEntropy-> HalfBinomialLoss
>    CategoricalCrossEntropy -> HalfMultinomialLoss
> I think this makes the names more consistent. If you dislike it, it's easy to revert that commit.

I think it's ok but it's important to document the alternative names in the docstring to make them grep'able / googl'able because the alternative names are very common in other ML frameworks (e.g. tensorflow / keras / pytorch).

ogrisel commented:
>  Is this a problem of this PR or can we ignore it? We don't know the deeper reason for it.

I don't think we can ignore it. Why would the code on this PR require data to be memory aligned?

> Should we disable readonly arrays for the losses as long as we can't fix this test?

We need support for readonly arrays because joblib memmaping is automatically used by `joblib.Parallel` to avoid large per-work processes memory copies of identical data buffers.

> How to proceed with the test? Just remove the part with the readonly array? Or use numpy.memmap directly which might work.

We can do that as a short-term workaround but we need to understand the root cause of the problem and either fix it in the Cython code of this PR or in joblib (but this sounds complicated).

ogrisel commented:
> @TomDLT Thanks for helping. If your guess is correct, do you have an explanation why it only happens on Ubuntu_Bionic py37_conda_forge_openblas_ubuntu_1804?

Maybe it's related to the version of the compiler used on that platform. Has someone manage to reproduce the segfault locally?

lorentzenchr commented:
> > Is this a problem of this PR or can we ignore it? We don't know the deeper reason for it.
>
> I don't think we can ignore it. Why would the code on this PR require data to be memory aligned?

The code requires contiguous 1-d memoryviews. Nothing spectacular.

> > I dared to rename the classification losses in 00328b2
> > - BinaryCrossEntropy-> HalfBinomialLoss
> > -  CategoricalCrossEntropy -> HalfMultinomialLoss
>
> I think it's ok but it's important to document the alternative names in the docstring to make them grep'able / googl'able because the alternative names are very common in other ML frameworks (e.g. tensorflow / keras / pytorch).

Yes, I put the alternative names on every occasion. I see my own background more in statistics and therefore prefer those names, and I'm writing this PR:wink: In the end, it's a private module, not a public API.

ogrisel commented:
I will try to debug further with the minimal reproducer and the bionic CI in #21677.

lorentzenchr committed:
Merge branch 'main' into loss_module_tempita

jjerphan commented:
@lorentzenchr:  A few fixes have been integrated in `main`; can you merge it back in your branch?

lorentzenchr committed:
Merge branch 'main' into loss_module_tempita

lorentzenchr committed:
trigger CI

lorentzenchr commented:
No luck anymore: CI failures on azure in the ccache step.
```
##[error]CaptureResult has already been called for action PipelineCache.RestoreCache
```

jjerphan commented:
:fire_extinguisher:  https://github.com/scikit-learn/scikit-learn/pull/21721

lorentzenchr committed:
Merge branch 'main' into loss_module_tempita

lorentzenchr committed:
TST aligned create_memmap_backed_data

lorentzenchr commented:
All 🟢 again!!!

lorentzenchr committed:
FIX replace CyBinaryCrossEntropy by CyHalfBinomialLoss

lorentzenchr committed:
MNT remove Cython compiler directives due to #21512

lorentzenchr commented:
Technically, there are 2 reviewers' approval. @thomasjpfan If you are also good with it, we could be close to merge?

BWT, thanks for convincing me of [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance). I should have read Design Patterns (1994) by the GoF more carefully. In particular, we avoid to rely on Cython multiple inheritance.

jjerphan reviewed (approved):
Second seal of approval since my last review: this LVGTM! Thank you, @lorentzenchr.

I think this PR has been standing still without attention for a long time.
I believe this can be merged but I would for other reviewers' approval.

rth commented:
> I believe this can be merged but I would for other reviewers' approval.

+1 for merging with 2 approvals, particularly since it doesn't affect any of the estimators yet and is a private API.

lorentzenchr commented:
> > I believe this can be merged but I would for other reviewers' approval.
> 
> +1 for merging with 2 approvals, particularly since it doesn't affect any of the estimators yet and is a private API.

If only someone would press that button 🙏 Preferably, it's not the committer of the PR.

lorentzenchr commented:
@rth ❤️
